PetProject/Services/CheapSharkService/CheapSharkService.swift

```diff
@@ -5,121 +5,107 @@
 //  Created by Тимофей Лукашевич on 29.03.22.
 //
 
-import Foundation
 import Combine
 import UIKit
 
 final class CheapSharkService {
     
     private let apiURL = URL(string: "https://www.cheapshark.com/api/1.0")
     private var hostURL =  URL(string: "https://www.cheapshark.com")
-    private var cancellables = Set<AnyCancellable>()
     private let networkService: NetworkService
-    let maximumDealsCount = 999
     let paginationDealsCount = 40
     
     init(networkService: NetworkService) {
         self.networkService = networkService
     }
     
-    func getStoresList(_ completion: @escaping (Result<[StoreModel], CheapSharkServiceError>) -> Void ) {
+    typealias GetStoresListPublisher = Publishers.Map<NetworkService.GetResponsePublisher, Result<[StoreModel], CheapSharkServiceError>>
+    func getStoresList() -> GetStoresListPublisher? {
         guard var url = apiURL else {
-            completion(.failure(.apiNotFound))
-            return
+            return nil
         }
         url.appendPathComponent("stores")
-        networkService.getResponsePublisher(url)
-            .timeout(3, scheduler: RunLoop.current)
-            .sink {
-                switch $0 {
-                case .failure(_):
-                    completion(.failure(.fetchError))
-                case .finished:
-                    print ("Received store list completion: \($0).")
-                }
-            } receiveValue: { (data: Data, response: URLResponse) in
-                if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
-                    do {
-                        let stores = try JSONDecoder().decode([StoreModel].self, from: data)
-                        completion(.success(stores))
-                    } catch let error {
-                        completion(.failure(.decodeError(error.localizedDescription)))
-                    }
-                } else {
-                    completion(.failure(.fetchError))
+        return networkService.getResponsePublisher(url).map { result -> Result<[StoreModel], CheapSharkServiceError> in
+            switch result {
+            case .failure(_):
+                return .failure(CheapSharkServiceError.fetchError)
+            case .success(let data):
+                do {
+                    let stores = try JSONDecoder().decode([StoreModel].self, from: data)
```

Comment:
So as not to constantly use decoding here. You can move it to the layer below, while using generic to get the desired type.
The standard case: make a request, get a response, check it and decode it. Return the expected model as a generic.

Reply:
Done